external-crates/move/crates/move-analyzer/src/analysis/typing_analysis.rs

```diff
@@ -437,6 +542,105 @@ impl TypingAnalysisContext<'_> {
         }
     }
+
+    fn process_match_patterm(&mut self, match_pat: &mut T::MatchPattern) {
+        use T::UnannotatedPat_ as UA;
+
+        self.visit_type(None, &mut match_pat.ty);
+        match &mut match_pat.pat.value {
+            UA::Variant(mident, name, vname, tyargs, fields)
+            | UA::BorrowVariant(_, mident, name, vname, tyargs, fields) => {
+                self.add_datatype_use_def(mident, name);
+                self.add_variant_use_def(mident, name, vname);
+                tyargs.iter_mut().for_each(|t| self.visit_type(None, t));
+                for (fpos, fname, (_, (_, pat))) in fields.iter_mut() {
+                    if self.compiler_info.ellipsis_binders.get(&fpos).is_none() {
+                        self.add_variant_field_use_def(
+                            &mident.value,
+                            &name.value(),
+                            &vname.value(),
+                            fname,
+                            &fpos,
+                        );
+                    }
+                    self.process_match_patterm(pat);
+                }
+            }
+            UA::Struct(mident, name, tyargs, fields)
+            | UA::BorrowStruct(_, mident, name, tyargs, fields) => {
+                self.add_datatype_use_def(mident, name);
+                tyargs.iter_mut().for_each(|t| self.visit_type(None, t));
+                for (fpos, fname, (_, (_, pat))) in fields.iter_mut() {
+                    if self.compiler_info.ellipsis_binders.get(&fpos).is_none() {
+                        self.add_struct_field_use_def(&mident.value, &name.value(), fname, &fpos);
+                    }
+                    self.process_match_patterm(pat);
+                }
+            }
+            UA::Constant(mod_ident, name) => self.add_const_use_def(mod_ident, name),
+            UA::Binder(mut_, var) => self.add_local_def(
+                &var.loc,
+                &var.value.name,
+                match_pat.ty.clone(),
+                false,
+                matches!(mut_, E::Mutability::Mut(_)),
+                false,
+            ),
+            UA::Or(pat1, pat2) => {
+                self.process_match_patterm(pat1);
+                self.process_match_patterm(pat2);
+            }
+            UA::At(var, pat) => {
+                self.add_local_def(
+                    &var.loc,
+                    &var.value.name,
+                    match_pat.ty.clone(),
+                    false,
+                    false,
+                    false,
+                );
+                self.process_match_patterm(pat);
+            }
+            UA::Literal(_) | UA::Wildcard | UA::ErrorPat => (),
+        }
+    }
+
+    fn process_match_arm(&mut self, sp!(_, arm): &mut T::MatchArm) {
+        self.process_match_patterm(&mut arm.pattern);
+        arm.binders.iter_mut().for_each(|(var, ty)| {
+            self.add_local_def(&var.loc, &var.value.name, ty.clone(), false, false, true);
+        });
```

Comment:
Do we need to add the locals for the patterns during pattern processing if we're also doing it here?